ECDH-ES Key Exchange Provider

cesarpayan opened the pull request:
Creating PR for feedback. PR will be 'blocked' until perf/stress testing is complete.

brentschmaltz committed:
Rebase and squash for EcdhEs

Creating EcdhKeyExchangeProvider
Adding ECDH-ES logic/cases to JsonWebTokenHandler
Adding target for net core 3.1

brentschmaltz committed:
Revert "Remove deprecated "language" attribute"

This reverts commit 5c51220085ffa4452d46799d1bae8a3a6bf0f8a8.

Adding immediate retry on network failure + better logging during configuation retrieval (#1784)

* Adding immediate retry on network failure to the token validation flow using ConfigurationManager (specifically inside of HttpDocumentRetriever)

* Adding better logging during configuration retrieval

Adding more information on key location to error messages (#1786)

rename JwtHandler -> JsonWebTokenHandler
adjust tests to throw invalid signature

Simplify the EventBasedLRUCache and Allows Skipping LRU (#1783)

* simplify the EventBasedLRUCache bit and added the _maintainLRU flag to skip the maintenance of LRU

* resolved review comments

* 1. let the event queue task continue to run for 2 min after the queue is empty 2. stop the task when the InMemoryCryptoProviderCache.Dispose() is called

* skip all operations on _doubleLinkedList when _maintainLRU = false

* fixed the failed test MaintainLRUOrder (_maintainLRU needs to set to true)

Update DisposableObjectPool to dispose on Free() when full

When the internal `items` array of DisposableObjectPool is full, calls to Free() (that are trying to return over-allocated instances during a spike in calls to Allocate() "drop" the object by doing nothing.

Since the object is not disposed, before it can be garbage collected it has to wait in the finalization queue for finalization to call dispose.

This change updates Free() to directly dispose those objects which can't be returned to the pool, allowing them to avoid a potentially long wait in the finalizer queue.

param check for null

fix check for tenantId (#1801)

update version to 6.15.2

Adding LKG feature into JwtSecurityTokenHandler

Adding tests for SignatureValidatorUsingConfiguration

address comment

add IsRecoverableConfiguration

mark metadata address as non-PII

Make M.IM.Tokens visible to S2S.Tokens (#1807)

added the ValidateTokenAsyc() and ReadToken() methods to all token handlers (#1810)

* added the ValidateTokenAsyc() method to all token handlers

* implement the ValidateTokenAsync() for JwtSecurityTokenHandler

* added the TokenHandler.ReadToken() method

* return async result via .ConfigureAwait(false).GetAwaiter().GetResult(), and don't catch general Exception

* added expected exceptions to the TokenHandler.cs() and removed CA1031 from GlobalSuppression as we are now caching specific exceptions

* added more comments

* updated all token handlers to catch the general exception (like in JsonWebTokenHandler) to be consistent

* updated comments

* always return the first ClaimsIdentity from the ClaimsPrincipal as TokenValidationResult.ClaimsIdentity

* return the first identity from the ClaimsPrincipal in Saml2SecurityTokenHandler.ValidateTokenAsync() so it is consistent with Saml1

update version to 6.16.1 (#1811)

update patch version after release of 6.16.0

Configuration validator (#1825)

Introduce an ConfigurationValidator class that can be used to apply validation rules to a retrieved configuration.

Simplify strings comparison with Ordinal option

update version

update version for next release

Add 'cty' claim to JWE header

Address comments

Fix tests (#1838)

Fix DEF test (#1839)

Copying work from broken topic branch
EcdsaKeyWrap Provider
Constants
Ecdh test

Creating EcdhKeyExchangeProvider
Removed unnecesary code and comments from ReferenceTests

Created Ecdh Security Key

Removing EcdhSecurityKey since ECDsaSecurityKey is the same thing
Modifying ctor for EcdhKeyExchangeProvider

Refactoring + overloads for ctrs in KeyExchangeProvider
Differentiate in between ECDH-ES and ECDH-ES+A{128|192|256}KW. Setting AlgorithmID accordingly to alg and enc values.
Returning CEK as SecurityKey
Changes to Reference test to reflect the changes

enc is not optional for key exchange provider
apu and apv can be null/empty for generate cek

added comments to guide ECDH-ES

Adding Jwt header params for epk, apu and apv
Adding supported algorithms for symmetric keywrap
Adding ECDH-ES logic/cases to JsonWebTokenHandler

added test for creating a JWE

Test cases for Jwe using Ecdh-ES
Reference test clean up
Use of apv and apu included
Aes192KW inclusion

LogMessages update in EcdhEs Key Exchange Provider
Removed unsued ctors
Added direct tests for EcdhEsKeyExchangeProvider
Cleaned up reference test

Refactoring ctors in EcdhEsKeyExchangeProvider, including new LogMessage
Removed unused file that was part of an earlier commit

Addressing feed back from PR
EcdhKeyExchangeProvider's constructor refactoring: Params for keys are both SecurityKeys, new error message to better describe when we were unable to obtain ECParameters, and some refactoring for readbility.
Refactoring for JsonWebTokenHandler.ResolveTokenDecryptionKey for string comparison
JwtTokenUtulities.GetSecurityKey refactor for readbility, null clauses for apu amd apv being null, and changed the creating of SymmetricSecurityKey to match with what is already in place for difference cases.
Test case scenario for the metnioned above.
Removed unnecesary code

Adding target for net core 3.1

Copying work from broken topic branch
EcdsaKeyWrap Provider
Constants
Ecdh test

Creating EcdhKeyExchangeProvider
Removed unnecesary code and comments from ReferenceTests

Created Ecdh Security Key

Removing EcdhSecurityKey since ECDsaSecurityKey is the same thing
Modifying ctor for EcdhKeyExchangeProvider

Refactoring + overloads for ctrs in KeyExchangeProvider
Differentiate in between ECDH-ES and ECDH-ES+A{128|192|256}KW. Setting AlgorithmID accordingly to alg and enc values.
Returning CEK as SecurityKey
Changes to Reference test to reflect the changes

enc is not optional for key exchange provider
apu and apv can be null/empty for generate cek

added test for creating a JWE

Test cases for Jwe using Ecdh-ES
Reference test clean up
Use of apv and apu included
Aes192KW inclusion

LogMessages update in EcdhEs Key Exchange Provider
Removed unsued ctors
Added direct tests for EcdhEsKeyExchangeProvider
Cleaned up reference test

Refactoring ctors in EcdhEsKeyExchangeProvider, including new LogMessage
Removed unused file that was part of an earlier commit

Addressing feed back from PR
EcdhKeyExchangeProvider's constructor refactoring: Params for keys are both SecurityKeys, new error message to better describe when we were unable to obtain ECParameters, and some refactoring for readbility.
Refactoring for JsonWebTokenHandler.ResolveTokenDecryptionKey for string comparison
JwtTokenUtulities.GetSecurityKey refactor for readbility, null clauses for apu amd apv being null, and changed the creating of SymmetricSecurityKey to match with what is already in place for difference cases.
Test case scenario for the metnioned above.
Removed unnecesary code

Adding target for net core 3.1

removing extra letter from rebase

Adding NET_CORE defined when using netcore 3.1

brentschmaltz commented on src/Microsoft.IdentityModel.Tokens/Encryption/EcdhKeyExchangeProvider.cs:
```diff
@@ -0,0 +1,275 @@
+        {
+            if (privateKey == null)
+                throw LogHelper.LogArgumentNullException(nameof(privateKey));
+            if (publicKey is null)
```
we should have a new-line separating clauses.

cesarpayan replied:
Fixed in a52ecf360f5d0463192c760898765e2a8542ee68

brentschmaltz commented on src/Microsoft.IdentityModel.Tokens/Encryption/EcdhKeyExchangeProvider.cs:
```diff
@@ -0,0 +1,275 @@
+        /// <param name="alg">alg header parameter value.</param>
+        /// <param name="enc">enc header parameter value.</param>
+        /// </summary>
+        public EcdhKeyExchangeProvider(ECDsaSecurityKey privateKey, ECDsaSecurityKey publicKey, string alg, string enc)
```
These two constructors do very similar work.
We could do two things
1. have a single ctor that takes a SecurityKey, then check that the keys are ECDsaSecurityKey OR a JsonWebKey that can be converted to a ECDsaSecurityKey.
2. have initialize methods that perform the common work.

cesarpayan replied:
Changed in a52ecf360f5d0463192c760898765e2a8542ee68
Taking the first suggestion

brentschmaltz commented on src/Microsoft.IdentityModel.Tokens/Encryption/EcdhKeyExchangeProvider.cs:
```diff
@@ -0,0 +1,275 @@
+            //The "apu" and "apv" values MUST be distinct when used (per rfc7518 section 4.6.2) https://datatracker.ietf.org/doc/html/rfc7518#section-4.6.2
+            if (!string.IsNullOrEmpty(apu)
+                && !string.IsNullOrEmpty(apv)
+                && apu.Equals(apv, StringComparison.InvariantCulture))
```
We should not specify culture here, we want OrdinalCaseSensitive, which is the default.

cesarpayan replied:
Fixed in a52ecf360f5d0463192c760898765e2a8542ee68

brentschmaltz commented on src/Microsoft.IdentityModel.Tokens/Encryption/EcdhKeyExchangeProvider.cs:
```diff
@@ -0,0 +1,275 @@
+
+            byte[] output = new byte[length];
+            int x = 0;
+            foreach (byte[] arr in arrays)
```
Can we use ArrayCopy as we do here: https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/039d09b7bc4c1671e1c4093d5a62d307c2f96935/src/Microsoft.IdentityModel.Tokens/Encryption/SymmetricKeyWrapProvider.cs#L280

cesarpayan replied:
Changed in a52ecf360f5d0463192c760898765e2a8542ee68

brentschmaltz commented on src/Microsoft.IdentityModel.Tokens/Encryption/EcdhKeyExchangeProvider.cs:
```diff
@@ -0,0 +1,275 @@
+
+namespace Microsoft.IdentityModel.Tokens
+{
+#if NET472
```
We need to add an additional target that can be used from .Net core.
We could add 5.0 and close this issue.
https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/issues/1426

cesarpayan replied:
Question: Should I make the additional target part of this PR?

Comment:
There are two referenced classes that have potential conflicts with versioning in this file:
- [ECParameters](https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.ecparameters?view=net-6.0) is available in all Core versions, and .NET Framework 4.7 onwards
- [ECDiffieHellman](https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.ecdiffiehellman?view=net-6.0) is available in Core 2.1 onwards, and .NET Framework 3.5 onwards

Would it look something like?
`#if NET472_OR_GREATER || NETSTANDARD2_1_OR_GREATER`

cesarpayan replied:
Per PR Meeting, we'll add .NET Core 3.1

cesarpayan replied:
Added in [f6f8794](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1834/commits/f6f8794f6309d9962e52064dd47858458ee10130)

cesarpayan reviewed (commented):
Addressing comments

brentschmaltz commented on src/Microsoft.IdentityModel.Tokens/Encryption/EcdhKeyExchangeProvider.cs:
```diff
@@ -56,51 +52,39 @@ public class EcdhKeyExchangeProvider //todo: rename to KeyExchangeProvider, or E
         /// <param name="enc">enc header parameter value.</param>
         /// </summary>
-        public EcdhKeyExchangeProvider(ECDsaSecurityKey privateKey, ECDsaSecurityKey publicKey, string alg, string enc)
+        public EcdhKeyExchangeProvider(ECDsaSecurityKey privateKey, SecurityKey publicKey, string alg, string enc)
```
An ECDsaSecurityKey private key could be represented by a JsonWebKey, unlikely since JsonWebKey is normally associated with public metadata.
That said, it is possible so both parameters could be SecurityKey.

cesarpayan replied:
Changed to SecurityKey and necessary changes below to accommodate to it in [1595e40](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1834/commits/1595e4097fc1b5c0f5b7a00fccf7f878bdd9c202)

brentschmaltz commented on src/Microsoft.IdentityModel.Tokens/LogMessages.cs:
```diff
@@ -263,6 +263,7 @@ internal static class LogMessages
         // Crypto Errors
         public const string IDX11000 = "IDX11000: Cannot create EcdhKeyExchangeProvider. '{0}'\'s Curve '{1}' does not match with '{2}'\'s curve '{3}'.";
         public const string IDX11001 = "IDX11001: Cannot generate KDF. '{0}':'{1}' and '{2}':'{3}' must be different.";
```
Perhaps a new error message for IDX1101 since the error is in the ctor:
```cs
public const string IDX11003 = "IDX11003: Cannot create the EcdhKeyExchangeProvider. the 'publicKey' cannot be converted to an ECDsaSecurityKey. publicKey type is: '{0}'.";
```

cesarpayan replied:
Since I'm removing the code path that lead to IDX1102, I am modifying the error message and using IDX1102 for that error case. Changes in [1595e40](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1834/commits/1595e4097fc1b5c0f5b7a00fccf7f878bdd9c202)

brentschmaltz commented on src/Microsoft.IdentityModel.Tokens/Encryption/EcdhKeyExchangeProvider.cs:
```diff
@@ -56,51 +52,39 @@ public class EcdhKeyExchangeProvider //todo: rename to KeyExchangeProvider, or E
                 throw LogHelper.LogArgumentNullException(nameof(publicKey));
 
-            ValidateAlgAndEnc(alg, enc);
+            if (publicKey is ECDsaSecurityKey ecdsaKey)
```
You should call [TryConvertToECDsaSecurityKey](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/050297acc5b1dacba8f706459af6da8803762453/src/Microsoft.IdentityModel.Tokens/JsonWebKeyConverter.cs#L344)

The reason is related to we have some logic in [ECDsaAdapter ](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/050297acc5b1dacba8f706459af6da8803762453/src/Microsoft.IdentityModel.Tokens/ECDsaAdapter.cs#L73)that can create an ECDsa object by examining the runtime, regardless of the 'target' specified by the application. Even though an application has targeted say .NET462, the actual runtime is most likely .NET48

You could write this as:
```cs
if (!TryConvertToECDsaSecurityKey(publicKey, out ecdaPublickey)
    throw LogHelper.LogArgumentException<ArgumentException>(
                    nameof(publicKey),
                    LogHelper.FormatInvariant(LogMessages.IDX11003, LogHelper.MarkAsNonPII(publicKey.GetType())));

    // then see if this works... if so, then GetEcParamsFromJwk would not be needed.
    _ecParamsPublic = ecdsaPublickey.ECDSA.ExportParameters(false);
```

cesarpayan replied:
I am refactoring this code in [1595e40](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1834/commits/1595e4097fc1b5c0f5b7a00fccf7f878bdd9c202) as now the ctor receives both keys as security keys as suggested above, to avoid duplicate code.
Adding it with log Message IDX11002 (previous id had been removed as I removed the code path that lead to it).
I am still checking if the security key is ECDsaSecurityKey or JsonWebKey as TryConvertToECDsaSecurityKey takes in a JsonWebKey.

brentschmaltz commented on src/Microsoft.IdentityModel.Tokens/Encryption/EcdhKeyExchangeProvider.cs:
```diff
@@ -56,51 +52,39 @@ public class EcdhKeyExchangeProvider //todo: rename to KeyExchangeProvider, or E
+            {
+                throw LogHelper.LogArgumentException<ArgumentException>(
+                    nameof(publicKey),
+                    LogHelper.FormatInvariant(LogMessages.IDX11001, LogHelper.MarkAsNonPII(nameof(publicKey))));
```
change error message, see comment in LogMessages.cs.

cesarpayan replied:
Changed to IDX11002 in [1595e40](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1834/commits/1595e4097fc1b5c0f5b7a00fccf7f878bdd9c202)

brentschmaltz commented on src/Microsoft.IdentityModel.Tokens/Encryption/EcdhKeyExchangeProvider.cs:
```diff
@@ -116,7 +100,7 @@ public SecurityKey GenerateKdf(string apu = null, string apv = null)
             if (!string.IsNullOrEmpty(apu)
                 && !string.IsNullOrEmpty(apv)
-                && apu.Equals(apv, StringComparison.InvariantCulture))
+                && apu.Equals(apv, StringComparison.Ordinal))
```
We found out that adding 'StringComparison.Ordinal', adds a performance penalty.
We should just leave it out.

cesarpayan replied:
Removed StringComparison.Ordinal and added a supression message with the performance penalty as justification in [1595e40](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1834/commits/1595e4097fc1b5c0f5b7a00fccf7f878bdd9c202)

dannybtsai commented on src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs:
```diff
@@ -755,13 +755,14 @@ private static string EncryptTokenPrivate(string innerJwt, EncryptingCredentials
                 throw LogHelper.LogExceptionMessage(new ArgumentException(TokenLogMessages.IDX10620));
 
```
This is a copy from the JwtSecurityTokenHandler.EncryptToken() but missing a null check on encryptingCredentials.
Suggest adding the GetCryptoProviderFactory() to the base class TokenHandler:

```
        protected static CryptoProviderFactory GetCryptoProviderFactory(EncryptingCredentials encryptingCredentials)
        {
            if (encryptingCredentials == null)
                throw LogHelper.LogArgumentNullException(nameof(encryptingCredentials));

            var cryptoProviderFactory = encryptingCredentials.CryptoProviderFactory ?? encryptingCredentials.Key.CryptoProviderFactory;

            if (cryptoProviderFactory == null)
                throw LogHelper.LogExceptionMessage(new ArgumentException(TokenLogMessages.IDX10620));

            return cryptoProviderFactory;
        }
```

And both JsonWebTokenHandler and JwtSecurityTokenHandler get the factory in a consistent manner:

`           var cryptoProviderFactory = GetCryptoProviderFactory(encryptingCredentials);
`

cesarpayan replied:
After inspecting the code paths into these two methods, I've realized that in neither case it is possible for these two methods to receive encyptingCredentials with a null value (each code path has a previous check for null value. As these two methods are private, I'll remove the check for null value in JwtSecurityTokenHandler.EncryptToken()

dannybtsai commented on src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs:
```diff
@@ -830,6 +835,24 @@ internal IEnumerable<SecurityKey> GetContentEncryptionKeys(JsonWebToken jwtToken
+                    }
+                    else
+#endif
                     if (key.CryptoProviderFactory.IsSupportedAlgorithm(jwtToken.Alg, key))
```
>CryptoProviderFactory [](http://example.com/codeflow?start=28&length=21)

Could this be null?

cesarpayan replied:
AFAIK, no. I don't think it is possible to have a SecurityKey with CryptoProviderFactory being null (there's a default being assigned when one isn't provided)

brentschmaltz replied:
SecurityKey.CryptoProviderFactory cannot be null.

dannybtsai commented on src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs:
```diff
@@ -886,7 +916,7 @@ protected virtual SecurityKey ResolveTokenDecryptionKey(string token, JsonWebTok
                 if (validationParameters.TokenDecryptionKey != null)
                 {
-                    if (string.Equals(validationParameters.TokenDecryptionKey.KeyId, jwtToken.X5t, validationParameters.TokenDecryptionKey is X509SecurityKey ? StringComparison.OrdinalIgnoreCase : StringComparison.Ordinal))
+                    if (string.Equals(validationParameters.TokenDecryptionKey.KeyId, jwtToken.X5t, stringComparison))
```
>stringComparison [](http://example.com/codeflow?start=99&length=16)

This is different than the original logic, is that correct? And it is differnt than the key collection below.

cesarpayan replied:
Not really, it is the same logic (if validationParameters.TokenDecryptionKey is x509 it'll use StringComparison.OrdinalIgnoreCase. The change is that I am pre-calculating which string comparison it is being used in order to save extra lines of code, combined with using functions as you suggested above, it should be clearer. Changes in [1595e40](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1834/commits/1595e4097fc1b5c0f5b7a00fccf7f878bdd9c202)

dannybtsai commented on src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs:
```diff
@@ -866,17 +889,24 @@ protected virtual SecurityKey ResolveTokenDecryptionKey(string token, JsonWebTok
+            StringComparison stringComparison = (validationParameters.TokenDecryptionKey is X509SecurityKey
+                || validationParameters.TokenDecryptionKey is ECDsaSecurityKey)
+                ? StringComparison.OrdinalIgnoreCase : StringComparison.Ordinal;
             if (!string.IsNullOrEmpty(jwtToken.Kid))
```
This is repeated and I got lost easily (there are two versions), suggest replacing with calls like the following:

        private static StringComparison GetCaseRuleIfX509(SecurityKey key) => key is X509SecurityKey ? StringComparison.OrdinalIgnoreCase : StringComparison.Ordinal;
        private static StringComparison GetCaseRuleIfECDsa(SecurityKey key) => (key is X509SecurityKey || key is ECDsaSecurityKey) ? StringComparison.OrdinalIgnoreCase : StringComparison.Ordinal;

cesarpayan replied:
Adding it in [1595e40](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1834/commits/1595e4097fc1b5c0f5b7a00fccf7f878bdd9c202)

dannybtsai commented on src/Microsoft.IdentityModel.JsonWebTokens/JwtTokenUtilities.cs:
```diff
@@ -292,6 +301,26 @@ internal static SecurityKey GetSecurityKey(EncryptingCredentials encryptingCrede
+                        apu = objApu.ToString();
+                    if (additionalHeaderClaims.TryGetValue(JwtHeaderParameterNames.Apv, out object objApv))
+                        apv = objApv.ToString();
+                }
```
objApu and objApv can be null.

cesarpayan replied:
Added clause for null and a unit test case for that scenario in [1595e40](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1834/commits/1595e4097fc1b5c0f5b7a00fccf7f878bdd9c202)

dannybtsai commented on src/Microsoft.IdentityModel.JsonWebTokens/JwtTokenUtilities.cs:
```diff
@@ -292,6 +301,26 @@ internal static SecurityKey GetSecurityKey(EncryptingCredentials encryptingCrede
+                EcdhKeyExchangeProvider ecdhKeyExchangeProvider = new EcdhKeyExchangeProvider(encryptingCredentials.Key as ECDsaSecurityKey, encryptingCredentials.JsonWebKey, encryptingCredentials.Alg, encryptingCredentials.Enc);
+                SecurityKey kdf = ecdhKeyExchangeProvider.GenerateKdf(apu, apv);
+                kwProvider = cryptoProviderFactory.CreateKeyWrapProvider(kdf, ecdhKeyExchangeProvider.GetEncryptionAlgorithm());
+                SecurityKey cek = new SymmetricSecurityKey(JwtTokenUtilities.GenerateKeyBytes(ecdhKeyExchangeProvider.KeyDataLen * 2));
```
>ecdhKeyExchangeProvider.KeyDataLen * 2 [](http://example.com/codeflow?start=94&length=38)

If ecdhKeyExchangeProvider.KeyDataLen is size in bits, what is the logic of 2x? This seems like a magic number and is not clear.

cesarpayan replied:
That's for the size for the wrapped key. Changing it to match the style/logic for AesCbcHmac below in [1595e40](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1834/commits/1595e4097fc1b5c0f5b7a00fccf7f878bdd9c202)

dannybtsai commented on src/Microsoft.IdentityModel.Tokens/EncryptingCredentials.cs:
```diff
@@ -109,6 +109,11 @@ public string Enc
 
+        /// <summary>
+        /// 
+        /// </summary>
```
Missing comments.

cesarpayan replied:
Added in [1595e40](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1834/commits/1595e4097fc1b5c0f5b7a00fccf7f878bdd9c202)

dannybtsai commented on src/Microsoft.IdentityModel.Tokens/JsonWebKey.cs:
```diff
@@ -333,6 +333,32 @@ internal RSAParameters CreateRsaParameters()
+        {
+            //todo: update log msgs
+            if (string.IsNullOrEmpty(Crv))
+                throw LogHelper.LogExceptionMessage(new ArgumentException(LogHelper.FormatInvariant(LogMessages.IDX10700, LogHelper.MarkAsNonPII(_className), LogHelper.MarkAsNonPII("Crv"))));
```
>"Crv" [](http://example.com/codeflow?start=181&length=5)

nameof(Crv)

cesarpayan replied:
Removing this piece of code as is no longer used in [1595e40](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1834/commits/1595e4097fc1b5c0f5b7a00fccf7f878bdd9c202)

dannybtsai commented on src/Microsoft.IdentityModel.Tokens/JsonWebKey.cs:
```diff
@@ -333,6 +333,32 @@ internal RSAParameters CreateRsaParameters()
+                throw LogHelper.LogExceptionMessage(new ArgumentException(LogHelper.FormatInvariant(LogMessages.IDX10700, LogHelper.MarkAsNonPII(_className), LogHelper.MarkAsNonPII("Crv"))));
+
+            if (string.IsNullOrEmpty(X))
+                throw LogHelper.LogExceptionMessage(new ArgumentException(LogHelper.FormatInvariant(LogMessages.IDX10700, LogHelper.MarkAsNonPII(_className), LogHelper.MarkAsNonPII("X"))));
```
>"X" [](http://example.com/codeflow?start=181&length=3)

nameof(X)

cesarpayan replied:
Removing this piece of code as is no longer used in [1595e40](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1834/commits/1595e4097fc1b5c0f5b7a00fccf7f878bdd9c202)

dannybtsai commented on src/Microsoft.IdentityModel.Tokens/JsonWebKey.cs:
```diff
@@ -333,6 +333,32 @@ internal RSAParameters CreateRsaParameters()
+                throw LogHelper.LogExceptionMessage(new ArgumentException(LogHelper.FormatInvariant(LogMessages.IDX10700, LogHelper.MarkAsNonPII(_className), LogHelper.MarkAsNonPII("X"))));
+
+            if (string.IsNullOrEmpty(Y))
+                throw LogHelper.LogExceptionMessage(new ArgumentException(LogHelper.FormatInvariant(LogMessages.IDX10700, LogHelper.MarkAsNonPII(_className), LogHelper.MarkAsNonPII("Y"))));
```
>"Y" [](http://example.com/codeflow?start=181&length=3)

nameof(Y)

cesarpayan replied:
Removing this piece of code as is no longer used in [1595e40](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1834/commits/1595e4097fc1b5c0f5b7a00fccf7f878bdd9c202)

RojaEnnam committed:
Rebase and squash for EcdhEs

Creating EcdhKeyExchangeProvider
Adding ECDH-ES logic/cases to JsonWebTokenHandler
Adding target for net core 3.1

brentschmaltz commented on src/Microsoft.IdentityModel.JsonWebTokens/JwtTokenUtilities.cs:
```diff
@@ -292,6 +301,26 @@ internal static SecurityKey GetSecurityKey(EncryptingCredentials encryptingCrede
+                if (additionalHeaderClaims != null && additionalHeaderClaims.Count > 0)
+                {
+                    if (additionalHeaderClaims.TryGetValue(JwtHeaderParameterNames.Apu, out object objApu))
+                        apu = objApu.ToString();
```
add a new line after 312 to separate clauses.

cesarpayan replied:
Added in [1595e40](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1834/commits/1595e4097fc1b5c0f5b7a00fccf7f878bdd9c202)

brentschmaltz commented on src/Microsoft.IdentityModel.Tokens/EncryptingCredentials.cs:
```diff
@@ -109,6 +109,11 @@ public string Enc
+        /// <summary>
+        /// 
+        /// </summary>
+        public JsonWebKey JsonWebKey { get; set; }
```
For ECDH, this can be a JsonWebKey OR a ECDsaSecurityKey.
Suggest a property

```cs
public SecurityKey KeyExchangePublicKey {get;set;}
```

cesarpayan replied:
Changed in [1595e40](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1834/commits/1595e4097fc1b5c0f5b7a00fccf7f878bdd9c202)

brentschmaltz commented on src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs:
```diff
@@ -830,6 +842,24 @@ internal IEnumerable<SecurityKey> GetContentEncryptionKeys(JsonWebToken jwtToken
             {
                 try
                 {
+#if NET472
```
We need to add a CORE 3.1 target

cesarpayan replied:
Added in [f6f8794](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1834/commits/f6f8794f6309d9962e52064dd47858458ee10130)

cesarpayan committed:
Merge branch 'cesar/EcdhEs' of https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet into cesar/EcdhEs

cesarpayan committed:
Merge branch 'cesar/EcdhEs' of https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet into cesar/EcdhEs

cesarpayan committed:
Merge branch 'cesar/EcdhEs' of https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet into cesar/EcdhEs

cesarpayan committed:
rebase fix

jennyf19 committed:
log cert thumbprint (#1820)

* log cert thumbprint

* PR feedback

cesarpayan committed:
Fixing dupe code from rebase onto

cesarpayan committed:
Merge branch 'cesar/EcdhEs' of https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet into cesar/EcdhEs

sruke committed:
Introduced custom log level enum to remove dependency on System.Diagnostics.Tracing.EventLevel in IIdentityLogger (#1843)

brentschmaltz committed:
update version to 6.18.1

cesarpayan committed:
Merge branch 'cesar/EcdhEs' of https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet into cesar/EcdhEs
